### ISSUE TITLE
Rope performance is Really Bad

### ISSUE THREAD

**Issue by hawkw** [hawkw/an-editor#34]
even after #28, we get the following benchmark results: 
```
test tests::rope_add_1000      ... bench:  92,545,093 ns/iter (+/- 8,325,688)
test tests::rope_insert_1000   ... bench:  18,670,209 ns/iter (+/- 2,987,965)
test tests::string_add_1000    ... bench:     227,800 ns/iter (+/- 8,419)
test tests::string_insert_1000 ... bench:   3,262,551 ns/iter (+/- 313,452)
```

Time to find some hotspots in need of optimisation....

**Comment by hawkw** [comment]
I suspect we're spending a lot of time in `rebalance()`, and I'm going to see if rebalancing performance can be improved. Gonna try to do some profiling & see what happens.

**Comment by hawkw** [comment]
@cjm00, do you have any thoughts?

**Comment by hawkw** [comment]
According to some quick profiling I did in Xcode Instruments, in the `rope_add_1000` test, about 39% of our time is spent `clone()`ing and a whopping 49% is spent in `drop()`

**Comment by hawkw** [comment]
Oh, you know what, the `add` test is biased against us a bit, since the Rope `+` operator is the immutable append operator. So each test iteration we were spending a lot of time creating new `Rope`s, and then `drop()`ping them, which wasn't happening for the `String`

**Comment by hawkw** [comment]
A slightly fairer benchmark gets us the following results:
```
test tests::rope_add_1000      ... bench:  19,723,159 ns/iter (+/- 2,204,592)
test tests::rope_insert_1000   ... bench:  24,254,046 ns/iter (+/- 5,445,680)
test tests::string_add_1000    ... bench:     249,592 ns/iter (+/- 84,816)
test tests::string_insert_1000 ... bench:   3,272,854 ns/iter (+/- 672,410)

``` 

Performance still needs work, but it is not quite as tragic as I thought.

**Comment by hawkw** [comment]
<img width="804" alt="screen shot 2016-12-26 at 3 49 28 pm" src="https://cloud.githubusercontent.com/assets/2796466/21486469/4d25617c-cb83-11e6-9674-ff90780e4764.png">
so, uh,

**Comment by hawkw** [comment]
with rebalancing disabled:
```
test tests::rope_add_1000      ... bench:  12,757,218 ns/iter (+/- 2,015,672)
test tests::rope_insert_1000   ... bench:   8,874,948 ns/iter (+/- 1,053,272)
test tests::string_add_1000    ... bench:     249,911 ns/iter (+/- 7,707)
test tests::string_insert_1000 ... bench:   3,156,994 ns/iter (+/- 300,301)
```
of course, we should get some performant benefits from maintaining a balanced Rope, but it looks like our current (recursive) implementation is a time-suck.

**Comment by deepinthebuild** [comment]
Yikes. Unfortunately the internet here is less reliable than I'd like, but
we'll have to deal with this for sure.

**Comment by hawkw** [comment]
That's fine, it can wait 'til you get back. Thinking about going ahead & rewriting `Rope` to allocate all nodes in an `Arena`...

**Comment by hawkw** [comment]
(something like this binary tree impl: https://github.com/TeXitoi/benchmarksgame-rs/blob/master/src/binary_trees.rs)

**Comment by deepinthebuild** [comment]
That's probably a good place to start. You could look at how xi-editor
implements its Rope as well, I've been meaning to do that.

**Comment by hawkw** [comment]
Their Rope implementation (having peeked at the source just now) is _not_ based on `Arena` allocation; they're using a persistent immutable rope with `Rc`s.

**Comment by hawkw** [comment]
(which, as a side note, I still think is probably a much easier way to implement an edit history - implementing all editing commands so that they're undo-able seems Not Trivial)

**Comment by hawkw** [comment]
Also, an `Rc` rope would cut down on cloning & allocation a _lot_, since similar ropes can just share a majority of their nodes. Might result in a significant performance improvement. 

Of course, if we go that route, we can't provide an API with method signatures that match `String`,  but I'm beginning to question whether matching `String`'s API exactly is really the best option (see #23)...

**Comment by hawkw** [comment]
I'm going to try throwing together quick implementations of both an `Arena` rope and an `Rc` rope, and doing some more in depth benchmarking.

We may also want to look into [`servo/tendril`](https://github.com/servo/tendril) as a replacement for `String`s in our `Rope`. A `Tendril` is essentially a string, but it doesn't allocate for text less than 8 bytes long. This might shave off a lot of allocations in cases when the `Rope` is fragmented, especially like when appending a lot of short text segments.

**Comment by hawkw** [comment]
The [`rope-persist` branch](https://github.com/hawkw/an-editor/tree/rope-persist) adds a _very quick_ first pass on a Rope with `Rc`s.  Rebalancing is not implemented yet, and I think there's a lot of improvements that could be made to the _way_ it uses its `Rc`ed pointers. 

Performance results: 
```
test tests::rope_add_1000      ... bench:  12,023,458 ns/iter (+/- 1,973,965)
test tests::rope_insert_1000   ... bench:   9,433,005 ns/iter (+/- 2,317,771)
test tests::string_add_1000    ... bench:     245,912 ns/iter (+/- 81,211)
test tests::string_insert_1000 ... bench:   3,577,902 ns/iter (+/- 1,048,494)
```
Very similar to performance with mutable `Rope`s. It looks like `add`ing _may_ be slightly faster and `insert`ing may be slightly slower, but both are probably within the margin of error enough that we can basically say "performance is the same".

**Comment by hawkw** [comment]
Persistent rope performance with `tendril`s:
```
test tests::rope_add_1000      ... bench:  11,727,258 ns/iter (+/- 1,593,874)
test tests::rope_insert_1000   ... bench:   9,225,393 ns/iter (+/- 1,529,663)
test tests::string_add_1000    ... bench:     220,916 ns/iter (+/- 36,505)
test tests::string_insert_1000 ... bench:   3,112,728 ns/iter (+/- 201,573)
```

**Comment by hawkw** [comment]
`rope-persist` also adds support for thread-safe (atomic) ropes, using `Arc`s, or, if `tendril` is enabled, using atomic `Tendril`s:

atomic ropes with `Arc`:
```
test tests::rope_add_1000      ... bench:  12,565,666 ns/iter (+/- 2,003,767)
test tests::rope_insert_1000   ... bench:   9,886,233 ns/iter (+/- 1,433,699)
test tests::string_add_1000    ... bench:     239,453 ns/iter (+/- 25,660)
test tests::string_insert_1000 ... bench:   3,311,874 ns/iter (+/- 461,709)
```
atomic ropes with atomic `Tendril`s:
```
test tests::rope_add_1000      ... bench:  12,056,925 ns/iter (+/- 1,668,730)
test tests::rope_insert_1000   ... bench:   8,824,835 ns/iter (+/- 1,247,299)
test tests::string_add_1000    ... bench:     223,351 ns/iter (+/- 71,106)
test tests::string_insert_1000 ... bench:   3,117,460 ns/iter (+/- 452,039)
```

As above, `tendril` gives us very slightly better performance but not significantly.

**Comment by hawkw** [comment]
I'm wondering what the performance of a mutable `Arena`-based Rope with `Tendril`s would look like. I'll start on writing that either this evening or tomorrow.

**Comment by hawkw** [comment]
One concern with an `Arena` is that, as @twisted-pear pointed out, it'll start to leak memory if the arena-backed structure lives too long. A potential solution could be adding a free list of nodes that can be overwritten

**Comment by hawkw** [comment]
Okay, I've made some changes that _massively_ improve our performance - it turns out that I had overlooked something really embarrassingly obvious. 

We were caching the length on each node, but never actually using the cached length - we were calculating it each time. And even though calculating a node's length isn't a really slow operation, we use lengths enough that we were spending NEARLY FIFTY PERCENT of our total run time in calls to `Node.len()`. I knew this was gonna be something really embarrassing 😳

Anyway, commit 2bb1602fae37e8616dc695918eb4695d1d13fa94 fixes this.

 New performance numbers on `master` (with rebalancing still disabled), and after making some tweaks to benchmarks across the board: 
```
test tests::rope_add_1000      ... bench:     141,526 ns/iter (+/- 38,626)
test tests::rope_insert_1000   ... bench:     370,765 ns/iter (+/- 108,240)
test tests::string_add_1000    ... bench:      56,680 ns/iter (+/- 8,351)
test tests::string_insert_1000 ... bench:   3,450,232 ns/iter (+/- 665,061)
```

I think you'll agree that's much less tragically bad. 

Note that we out-perform `String` when inserting to a long string, as we ought to; and performance in the cases where we'd expect `String`s to be faster is no longer so much worse that it renders `Rope`s unusable.
